File: 775-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> children;

    public Node() {}

    public Node(int _val) {
        val = _val;
    }

    public Node(int _val,IList<Node> _children) {
        val = _val;
        children = _children;
    }
}
*/

public class Solution {

	List<int> ret = new List<int>();
    
	public IList<int> Preorder(Node root) {
		dopreorder(root);
		return ret;
    }
	
	public void dopreorder(Node root){
		if(root == null) return;
		ret.Add(root.val);
		foreach(Node node in root.children){
			dopreorder(node);
		}
	}
}

File: 775-n-ary-tree-preorder-traversal/n-ary-tree-preorder-traversal.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     int numChildren;
 *     struct Node** children;
 * };
 */

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
void order(struct Node* root, int* returnSize, int* res){
    if(!root) return;
    res[(*returnSize)++] = root->val;
    for(int i = 0; i < root->numChildren; i++){
        order(root->children[i], returnSize, res);
    }
}

int* preorder(struct Node* root, int* returnSize) {
    *returnSize = 0;
    int* res = (int*)malloc(sizeof(int) * 10000);
    order(root, returnSize, res);
    return res;
}